ext/semian/sysv_semaphores.c

```diff
@@ -14,28 +23,35 @@ raise_semian_syscall_error(const char *syscall, int error_num)
   rb_raise(eSyscall, "%s failed, errno: %d (%s)", syscall, error_num, strerror(error_num));
 }
 
-key_t
-generate_key(const char *name)
+int
+initialize_semaphore_set(const char* id_str, long permissions, int tickets, double quota)
 {
-  char semset_size_key[20];
-  char *uniq_id_str;
+  int sem_id;
+  key_t key;
 
-  // It is necessary for the cardinatily of the semaphore set to be part of the key
-  // or else sem_get will complain that we have requested an incorrect number of sems
-  // for the desired key, and have changed the number of semaphores for a given key
-  sprintf(semset_size_key, "_NUM_SEMS_%d", SI_NUM_SEMAPHORES);
-  uniq_id_str = malloc(strlen(name)+strlen(semset_size_key)+1);
-  strcpy(uniq_id_str, name);
-  strcat(uniq_id_str, semset_size_key);
+  key = generate_key(id_str);
+  sem_id = semget(key, SI_NUM_SEMAPHORES, IPC_CREAT | IPC_EXCL | permissions);
 
-  union {
-    unsigned char str[SHA_DIGEST_LENGTH];
-    key_t key;
-  } digest;
-  SHA1((const unsigned char *) uniq_id_str, strlen(uniq_id_str), digest.str);
-  free(uniq_id_str);
-  /* TODO: compile-time assertion that sizeof(key_t) > SHA_DIGEST_LENGTH */
-  return digest.key;
+  if (sem_id != -1) {
```

Comment:
Note: This has been based off of http://man7.org/tlpi/code/online/dist/svsem/svsem_good_init.c.html which avoids a race condition when initializing the semaphore set. Our previous implementation was pretty literally [how to do it the wrong way](http://man7.org/tlpi/code/online/diff/svsem/svsem_bad_init.c.html).

It's still possible that the process that tries to init will die, but in that case we will throw an error.

It's possible that we may want to fallback to the "Bad" approach, but I think an explicit error is better.